mark.c

```diff
@@ -810,23 +816,38 @@ GC_INNER mse * GC_mark_from(mse *mark_stack_top, mse *mark_stack,
         for(;;) {
           PREFETCH(limit - PREF_DIST*CACHE_LINE_SIZE);
           GC_ASSERT((word)limit >= (word)current_p);
-          deferred = *(word *)limit;
-          FIXUP_POINTER(deferred);
-          limit -= ALIGNMENT;
-          if (deferred >= (word)least_ha && deferred < (word)greatest_ha) {
-            PREFETCH((ptr_t)deferred);
-            break;
-          }
+          if( cheri_getaddress(limit) < cheri_getbase(limit)
```

Comment:
Formatting: `if( cheri...` should be `if (cheri...)` to fit in with the surroundings.

Reply:
I must admit I'm not quite sure I follow the logic in the conditional here: if `cheri_getaddress(limit) < cheri_getbase(limit)` do I want to do `limit -= ALIGNMENT;` or `limit += ALIGNMENT;`?

Reply:
fixed in commit [408f667b](https://github.com/capablevms/bdwgc/pull/1/commits/408f667bb59d4d389e44ff70d5631bf922413ed9)

Reply:
Can `SPLIT_RANGE_WORDS` be negative?

Reply:
SPLIT_RANGE_WORDS is  defined as 128 at the top of the function. I have not made any changes to it. I assume that is what the original authors intended.

Reply:
I am really struggling to follow the additional CHERI logic here. Can you maybe try expanding it into prose for me, bearing in mind that I'm an idiot, and need simple explanations?

Reply:
The additional CHERI logic is necessary to ensure that we only deference valid pointers while scanning. 

In non-CHERI code, the pointer would be de-referenced and the  conditional check immediately after would check whether the content is within a valid (non-blacklisted) heap address range.  If so then break out of the loop, push the contents for scanning and then continue marking. 

In CHERI-land, the pointer is only de-referenced if it is a valid pointer.  If it cannot be de-referenced for further marking, the code will skip that location and keep going.  Otherwise, it does what the original code does (i.e. check whether the de-referenced pointer is a plausible heap location).

Reply:
I think that part of my confusion is this part:

```c
if (cheri_getaddress(limit) < cheri_getbase(limit)) || ... ) limit -= ALIGNMENT;
```

Maybe I'm reading that incorrectly, but it seems to say "if the current address is below the base, subtract from the address", which seems backwards to me. I was expecting "if the current address is below the base, add to the address". I'm sure I'm wrong, but I'm struggling to tell how!

Reply:
I have a separate branch with a small optimization which I intend to submit if this one gets merged.
If the pointer is out of bounds then the loop can either be 

- sped up  --OR--
- short-circuited.

Reply:
At the moment I'm questioning the correctness of the change on this branch: the code doesn't seem correct to me, and I'm hoping you'll either explain to me why it's correct or fix it :) [I'm fine with a later PR optimising things, but I think each PR should (except in limited circumstances) be correct in and of itself.]

Reply:
The original code was already doing it this way  i.e. 
```
loop_top: 
limit = (current_p=get_next()) + CONST_SIZE
do 
 deref pointer limit;
 limit -= ALIGNMENT;
until ( current_p > limit ); 
goto loop_top;
```

The only change I have made in this PR, is to ensure that the pointer **can** be dereferenced before doing so ie.
```
loop_top: 
limit = (current_p=get_next())  + CONST_SIZE
do 
 if (guard with cheri conditions)
     deref pointer limit; 
 limit -= ALIGNMENT;
until ( current_p > limit ); 
goto loop_top;
```

This retains the current loop structure by which the pointer progresses within the  nested loops.

Reply:
Yes it's the "guard with CHERI conditions" bit that I'm unsure about. @jacobbramley had some thoughts offline: I wonder if he can repeat them here? Then we can start discussing the nitty gritty details :)

Reply:
As mentioned elsewhere:
1. The guard condition only examines the byte at the capability's address. Since a full word is accessed, the limit check should account for that. Otherwise, we'll erroneously accept accesses that span the limit. If the capability is known to be suitably aligned then this can't happen, but that isn't clear to me in this context. (The function is very long and has a lot of conditional blocks that I haven't entirely followed.)
2. (Resolved already, but I want to keep the numbers synchronised.)
3. I'd wrap that "guard with cheri condition" in a function. That would greatly clarify the code, and I'm sure it'll be useful elsewhere in the project at some point.

Reply:
Actually, the sole purpose of that loop is to prefetch an access that will be made later. I'd be tempted to just drop in a `FIXUP CHERI` comment and do it in a separate PR, but the comment at the top says that it's critical to performance so I expect we'd do it at some point anyway.

Another thought, after examining the context a bit more: `limit` is derived from `current_p`, and (for some reason I don't currently understand) is conveniently iterated downwards. We can, and should, use the limit of `current_p` to constrain the starting address of `limit`. This avoids the need to check the upper bound at all (though an assertion would be useful).

This is actually a possible correctness issue in the current version of the PR: if `limit` is initialised with a very large length (e.g. in `limit = current_p + <length>`) then it will have no tag, and the current "guard with cheri conditions" will catch it and bial out, when really it should have just been less eager. I don't know if that's likely to occur in practice.

The guard function (3) is probably not necessary, taking all that into account, since we only really need to examine the base.

> Maybe I'm reading that incorrectly, but it seems to say "if the current address is below the base, subtract from the address", which seems backwards to me. I was expecting "if the current address is below the base, add to the address". I'm sure I'm wrong, but I'm struggling to tell how!

@ltratt Actually I think it's _almost_ correct, but hard to read. The `limit` is decremented in both branches, so that condition is really just guarding the dereference, and the `[least_ha,greatest_ha)` check, which `break`s. The problem here is that it assumes that the `(word)current_p > (word)limit` bail-out check is triggered IFF the "guard with cheri conditions" fails, which is not necessarily the case, as far as I can tell.

Basically, what happens if `cheri_getbase(current_p) != current_p`?


This function is complicated and I'm sure that I've mis-analysed something and quite probably got the wrong idea somewhere. There are some subtleties here.

Reply:
Thanks @jacobbramley. I had not considered the spanning of boundaries at the limit. Had some of the changes on a separate branch, but moving it to this PR. Please could @ltratt, @jacobbramley and others check commit [9562b0d8](https://github.com/djichthys/bdwgc/commit/9562b0d8a9769f92b025ece1fbcf6c55747a2d6d). I have a gut feeling, I have missed something in it, but most of the checks are there I think.